lib/cn/csched_sp3.c

```diff
@@ -1592,37 +1592,48 @@ rest_csched_sp3_get(
     INVARIANT(resp);
     INVARIANT(arg);
 
-    status = REST_STATUS_INTERNAL_SERVER_ERROR;
     sp = arg;
 
     err = rest_params_get(req->rr_params, "pretty", &pretty, false);
-    if (err)
-        return REST_STATUS_BAD_REQUEST;
+    if (ev(err))
+        return rest_response_perror(resp, REST_STATUS_BAD_REQUEST,
+            "The 'pretty' query parameter must be a boolean");
 
     root = cJSON_CreateArray();
     if (ev(!root))
-        return REST_STATUS_INTERNAL_SERVER_ERROR;
+        return rest_response_perror(resp, REST_STATUS_SERVICE_UNAVAILABLE, "Out of memory");
 
     err = sts_foreach_job(sp->sts, sp3_job_serialize, root);
-    if (ev(err))
-        goto out;
+    if (ev(err)) {
+        switch (merr_errno(err)) {
+        case ENOMEM:
+            status = rest_response_perror(resp, REST_STATUS_SERVICE_UNAVAILABLE, "Out of memory");
+            goto out;
+        default:
+            status = rest_response_perror(resp, REST_STATUS_INTERNAL_SERVER_ERROR,
+                "Unhandled error");
+            goto out;
+        }
+    }
 
     str = (pretty ? cJSON_Print : cJSON_PrintUnformatted)(root);
-    if (!str)
+    if (ev(!str)) {
+        status = rest_response_perror(resp, REST_STATUS_SERVICE_UNAVAILABLE, "Out of memory");
         goto out;
+    }
 
     fputs(str, resp->rr_stream);
     cJSON_free(str);
 
-    err = rest_headers_set(resp->rr_headers, REST_HEADER_CONTENT_TYPE, REST_APPLICATION_JSON);
-    if (ev(err))
-        goto out;
-
+    rest_headers_set(resp->rr_headers, REST_HEADER_CONTENT_TYPE, REST_APPLICATION_JSON);
     status = REST_STATUS_OK;
 
 out:
     cJSON_Delete(root);
 
+    if (status != REST_STATUS_OK)
+        return rest_response_perror(resp, REST_STATUS_SERVICE_UNAVAILABLE, "Out of memory");
+
```

Comment:
I think it would be more useful if you just replaced all this error handling with a message that includes the decoded err.  Should be much less code and result in error messages that have some meat in them.